src/HDF5.jl

```diff
@@ -1040,34 +1054,26 @@ dataspace(attr::Attribute) = Dataspace(h5a_get_space(checkvalid(attr).id))
 
 # Create a dataspace from in-memory types
 dataspace(x::Union{T, Complex{T}}) where {T<:ScalarType} = Dataspace(h5s_create(H5S_SCALAR))
+dataspace(str::String) = Dataspace(h5s_create(H5S_SCALAR))
 
 function _dataspace(sz::Dims{N}, max_dims::Union{Dims{N}, Tuple{}}=()) where N
-    dims = Vector{hsize_t}(undef,length(sz))
-    any_zero = false
-    for i = 1:length(sz)
-        dims[end-i+1] = sz[i]
-        any_zero |= sz[i] == 0
-    end
-
-    if any_zero
-        space_id = h5s_create(H5S_NULL)
+    dims = hsize_t[sz[i] for i in N:-1:1]
+    if isempty(max_dims)
+        maxd = dims
     else
-        if isempty(max_dims)
-            maxd = dims
-        else
-            # This allows max_dims to be specified as -1 without triggering an overflow
-            # exception due to the signed -> unsigned conversion.
-            maxd = [(hssize_t.(reverse(max_dims)) .% hsize_t)...]
-        end
-        space_id = h5s_create_simple(length(dims), dims, maxd)
+        # This allows max_dims to be specified as -1 without triggering an overflow
+        # exception due to the signed -> unsigned conversion.
+        maxd = hsize_t[hssize_t(max_dims[i]) % hsize_t for i in N:-1:1]
     end
-    Dataspace(space_id)
+    return Dataspace(h5s_create_simple(length(dims), dims, maxd))
 end
-dataspace(A::AbstractArray; max_dims::Union{Dims,Tuple{}} = ()) = _dataspace(size(A), max_dims)
-dataspace(str::String) = Dataspace(h5s_create(H5S_SCALAR))
+dataspace(A::AbstractArray{T,N}; max_dims::Union{Dims{N},Tuple{}} = ()) where {T,N} = _dataspace(size(A), max_dims)
+# special array types
 dataspace(v::VLen; max_dims::Union{Dims,Tuple{}}=()) = _dataspace(size(v.data), max_dims)
+dataspace(A::EmptyArray) = Dataspace(h5s_create(H5S_NULL))
 dataspace(n::Nothing) = Dataspace(h5s_create(H5S_NULL))
-dataspace(sz::Dims; max_dims::Union{Dims,Tuple{}}=()) = _dataspace(sz, max_dims)
+# for giving sizes explicitly
+dataspace(sz::Dims{N}; max_dims::Union{Dims{N},Tuple{}}=()) where {N} = _dataspace(sz, max_dims)
```

Comment:
Is this and the parametric restriction above done for clarity/increased specificity (I don't think this will improve performance)?

Reply:
Yes, just for exactness, with no intention that it actually makes this more performant. The only valid options are for default/empty `max_dims` or for `length(sz) == length(max_dims)`, so we might as well enforce that in the type signature.